src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java

```diff
@@ -106,112 +126,104 @@ public int getMaxLength() {
 	public int getCursor() {
 		return this.cursor;
 	}
-	
+
+	@Environment(EnvType.CLIENT)
+	public void scrollCursorIntoView() {
+		if (scrollOffset > cursor) {
+			scrollOffset = cursor;
+		}
+		if (scrollOffset < cursor && font.trimToWidth(text.substring(scrollOffset), width - TEXT_PADDING_X * 2).length() + scrollOffset < cursor) {
+			scrollOffset = cursor;
+		}
+
+		int rightMostScrollOffset = text.length() - font.trimToWidth(text, width - TEXT_PADDING_X * 2, true).length();
+		scrollOffset = Math.min(rightMostScrollOffset, scrollOffset);
+	}
+
 	@Nullable
 	public String getSelection() {
-		if (select<0) return null;
-		if (select==cursor) return null;
-		
+		if (select < 0) return null;
+		if (select == cursor) return null;
+
 		//Tidy some things
-		if (select>text.length()) select = text.length();
-		if (cursor<0) cursor = 0;
-		if (cursor>text.length()) cursor = text.length();
-		
+		if (select > text.length()) select = text.length();
+		if (cursor < 0) cursor = 0;
+		if (cursor > text.length()) cursor = text.length();
+
 		int start = Math.min(select, cursor);
 		int end = Math.max(select, cursor);
-		
+
 		return text.substring(start, end);
 	}
-	
+
 	public boolean isEditable() {
 		return this.editable;
 	}
 
 	@Environment(EnvType.CLIENT)
-	protected void renderTextField(MatrixStack matrices, int x, int y) {
```

Comment:
Same here

Reply:
renderTextField still exists and just does more stuff now (e.g. render cursor blinking and selection and all that stuff i added.)

Reply:
Ah sorry, I didn't notice